libnavigation-core/src/test/java/com/mapbox/navigation/core/preview/RoutesPreviewControllerTest.kt

```diff
@@ -0,0 +1,404 @@
+            previewUpdate!!.routesPreview!!.alternativesMetadata.first().navigationRoute
+        )
+
+        val result = previewUpdate!!.routesPreview
+        assertNotNull(result)
+        result!!
+        assertEquals(testRoutes[1], result.primaryRoute)
+        assertEquals(testRoutes[0], result.alternativesMetadata.first().navigationRoute)
+        assertEquals(testRoutes, result.originalRoutesList)
+        assertEquals(
+            listOf(
+                testRoutes[1],
+                testRoutes[0]
+            ),
+            result.routesList
+        )
+        assertEquals(1, result.primaryRouteIndex)
+    }
+
+    @Test
+    fun `switch between previewed routes quicker then internal processing`() {
+        val testRoutes = createNavigationRoutes(
+            response = createDirectionsResponse(
+                uuid = "test",
+                routes = listOf(
+                    createDirectionsRoute(),
+                    createDirectionsRoute(),
+                    createDirectionsRoute(),
+                )
+            )
+        )
+        val waitHandle = CompletableDeferred<Unit>()
+        val routesPreviewController = createRoutePreviewController(
+            routesDataParser = {
+                if (it.first().id == "test#1") {
+                    waitHandle.await()
+                }
+                RouteDataParserStub().parse(it)
+            }
+        )
+        val previewedRoutesIds = mutableListOf<List<String>?>()
+        routesPreviewController.registerRoutesPreviewObserver {
+            previewedRoutesIds.add(it.routesPreview?.routesList?.map { it.id })
+        }
+        routesPreviewController.previewNavigationRoutes(testRoutes)
+
+        routesPreviewController.changeRoutesPreviewPrimaryRoute(testRoutes[1])
+        routesPreviewController.changeRoutesPreviewPrimaryRoute(testRoutes[2])
+        waitHandle.complete(Unit)
+
+        assertEquals(
+            listOf(
+                listOf("test#0", "test#1", "test#2"),
+                listOf("test#1", "test#0", "test#2"),
+                listOf("test#2", "test#0", "test#1")
+            ),
+            previewedRoutesIds
+        )
+    }
+
+    @Test(expected = IllegalArgumentException::class)
+    fun `select wrong primary route`() {
+        val routesPreviewController = createRoutePreviewController()
+        routesPreviewController.previewNavigationRoutes(
+            createNavigationRoutes(
+                response = createDirectionsResponse(
+                    uuid = "test",
+                    routes = listOf(
+                        createDirectionsRoute(),
+                        createDirectionsRoute()
+                    )
+                )
+            )
+        )
+
+        routesPreviewController.changeRoutesPreviewPrimaryRoute(
+            createNavigationRoutes().first()
+        )
+    }
+
+    @Test(expected = IllegalArgumentException::class)
+    fun `select primary route without previewing any routes`() {
```

Comment:
You can also check what happens when you invoke `changeRoutesPreviewPrimaryRoute` after clearing routesPreview.